CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+# CIP [0017]: Integrate Plumo Proofs into the lightest sync protocol
+
+- Date: 2020-07-27
+- Author: @lucasege
+- Status: WIP
+
+## Overview
+
+In order to continue supporting the dream of a seamless mobile sync experience, the lightest sync protocol must be extended to support PLUMO Zk-SNARKs which can quickly verify transitions from epoch X to epoch X + Y, where Y could be on the order of 10s-100s of epochs.
+
+The prover/verifier for plumo proofs have already been implemented, and the exposed go interface is available [here](https://github.com/celo-org/celo-bls-go)
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
+and
+`Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.
+
+This `AddProof` function will store the proof in a new database `plumoproofdata`, keyed by a unique proof ID.
+
+To propogate these proofs throughout the network, we also propose to extend the `eth` protocol with two new message types: 
+```
+PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`
+GetPlumoProofMsg - Request plumo proof(s) from a peer
+```
+`GetPlumoProofMsg` will be implemented by having the requester send a set of proof IDs that they hold, and ask for any they are missing. This may be optimized using a bloom filter as well.
+
+The `les` protocol must then also be extended via at least two messages:
+```
+GetEpochsMsg - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.
+EpochsMsg - Receives a mix of header(s) and proof(s) from the server.
```

Comment:
I would propose that that instead of this extension we add messages that will allow the client to be inform of what proofs the server can provide filtered by block range, and to download those proofs. That would give the client flexibility to decide what mix of proofs and headers it wants to use for syncing.

Reply:
hm - I think this is something worth discussing synchronously. I'm personally partial to the original plan, because it seems to reduce any necessary back/forth between client/server.
I don't see any uses for the client being able to chose the mix of proofs/headers, is there a benefit you see?
The motivation behind the original approach was that the light client doesn't care _how_ it syncs, only that it syncs in the fastest way. It seems logical imo for the server (with all relevant info, no need to download more) to interpret this and return the data necessary to make this happen